### PR TITLE
Failover marker persistence

### PR DESCRIPTION
**Pull request by yux0**
**What changed?**
Provide persistence API to insert failover marker to the replication queue

**Why?**
To support graceful failover

**How did you test it?**
Integration test

**Potential risks**
N/A

### CODE REVIEW

File: common/constants.go [hunk 1]
```diff
@@ -52,6 +52,8 @@ const (
 const (
 	// EmptyUUID is the placeholder for UUID when it's empty
 	EmptyUUID = "emptyUuid"
+	// EmptyReplicationUUID is the placeholder for replication task UUID when it's empty
+	EmptyReplicationUUID = "30000000-5000-f000-f000-000000000000"
```
Comment (yycptt) [inline common/constants.go:56]: I think this constant should live in `persistence` package

File: common/persistence/cassandra/cassandraPersistence.go [hunk 2]
```diff
@@ -2892,3 +2894,26 @@ func (d *cassandraPersistence) RangeDeleteReplicationTaskFromDLQ(
...
+			task.DomainID,
+			rowTypeReplicationWorkflowID,
+			rowTypeReplicationRunID,
+			task.GetVisibilityTimestamp().UnixNano(),
+		); err != nil {
+			return err
+		}
+	}
+
+	return d.session.ExecuteBatch(batch)
```
Comment (yycptt) [inline common/persistence/cassandra/cassandraPersistence.go:2918]: Please convert common gocql error to `ServiceBusyError`, `InternalServiceError` etc.

File: common/persistence/cassandra/cassandraPersistenceUtil.go [hunk 3]
```diff
@@ -1054,6 +1059,7 @@ func createReplicationTasks(
 	domainID string,
 	workflowID string,
 	runID string,
+	creationTime int64,
```
Comment (yycptt) [inline common/persistence/cassandra/cassandraPersistenceUtil.go:1062]: Seems like all replication task types already have the `VisibilityTimestamp` field. Can we reuse value from that field?

File: common/persistence/dataInterfaces.go [hunk 4]
```diff
@@ -609,9 +610,8 @@ type (
 	// FailoverMarkerTask is the marker for graceful failover
 	FailoverMarkerTask struct {
 		TaskID              int64
-		VisibilityTimestamp time.Time
+		VisibilityTimestamp time.Time // Visibility timestamp stores in the field scheduleEventID
```
Comment (yycptt) [inline common/persistence/dataInterfaces.go:613]: Please update the comment.

File: common/persistence/sql/sqlExecutionManager.go [hunk 5]
```diff
@@ -28,6 +28,8 @@ import (
 	"math"
 	"time"
 
+	"github.com/uber/cadence/common/log/tag"
+
```
Comment (yycptt) [inline common/persistence/sql/sqlExecutionManager.go:32]: nit: Please remove empty line.

File: common/persistence/sql/sqlExecutionManagerUtil.go [hunk 6]
```diff
@@ -880,6 +886,11 @@ func createReplicationTasks(
 			activityScheduleID = task.(*p.SyncActivityTask).ScheduledID
 			lastReplicationInfo = map[string]*sqlblobs.ReplicationInfo{}
 
+		case p.ReplicationTaskTypeFailoverMarker:
+			version = task.GetVersion()
+			// Failover marker uses firstEventID to store visibility timestamp
```
Comment (yycptt) [inline common/persistence/sql/sqlExecutionManagerUtil.go:891]: Please update the comment

File: common/persistence/cassandra/cassandraPersistence.go [hunk 7]
```diff
@@ -2892,3 +2894,26 @@ func (d *cassandraPersistence) RangeDeleteReplicationTaskFromDLQ(
...
 	return nil
 }
+
+func (d *cassandraPersistence) CreateFailoverMarkerTasks(
+	request *p.CreateFailoverMarkersRequest,
+) error {
+
+	batch := d.session.NewBatch(gocql.LoggedBatch)
+	for _, task := range request.Markers {
+		t := []p.Task{task}
```
Comment (yycptt) [inline common/persistence/cassandra/cassandraPersistence.go:2904]: I think I am missing some context here. `createReplicationTasks` can take a list of tasks, why do we need to invoke it for only one task at a time?

File: common/constants.go (hunk 1 above)
Comment (yux0) [reply common/constants.go]: Moved to sql package

File: common/persistence/cassandra/cassandraPersistence.go (hunk 2 above)
Comment (yux0) [reply common/persistence/cassandra/cassandraPersistence.go]: Handled the gocql errors

File: common/persistence/cassandra/cassandraPersistenceUtil.go (hunk 3 above)
Comment (yux0) [reply common/persistence/cassandra/cassandraPersistenceUtil.go:1062]: The VisibilityTimestamp lives in execution table. It uses a default timestamp value for insert and select as VisibilityTimestamp is a part of the index. So I cannot use the field

File: common/persistence/dataInterfaces.go (hunk 4 above)
Comment (yux0) [reply common/persistence/dataInterfaces.go]: Removed the comment

File: common/persistence/sql/sqlExecutionManagerUtil.go (hunk 6 above)
Comment (yux0) [reply common/persistence/sql/sqlExecutionManagerUtil.go]: Removed the comment

File: common/persistence/sql/sqlExecutionManager.go (hunk 5 above)
Comment (yux0) [reply common/persistence/sql/sqlExecutionManager.go]: Updated

File: common/persistence/cassandra/cassandraPersistence.go (hunk 7 above)
Comment (yux0) [reply common/persistence/cassandra/cassandraPersistence.go:2904]: The tasks have different domain ID so it has to update one at a time.

File: common/persistence/cassandra/cassandraPersistenceUtil.go (hunk 3 above)
Comment (yux0) [reply common/persistence/cassandra/cassandraPersistenceUtil.go]: Updated all replication task creation time to use VisibilityTimestamp

**Pull request merged**